src/transformers/models/llama/modeling_llama.py

```diff
@@ -590,15 +645,133 @@ def _upad_input(self, query_layer, key_layer, value_layer, padding_mask, query_l
         )
 
 
+class LlamaSDPAAttention(LlamaAttention):
+    """
+    Llama attention module using torch.nn.functional.scaled_dot_product_attention. This module inherits from
+    `LlamaAttention` as the weights of the module stays untouched. The only changes are on the forward pass to adapt to
+    SDPA API.
+    """
+
+    # Adapted from LlamaAttention.forward
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        padding_mask: Optional[torch.LongTensor] = None,
```

Comment:
```suggestion
        attention_mask: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.LongTensor] = None,
        past_key_value: Optional[Tuple[torch.Tensor]] = None,
        output_attentions: bool = False,
        use_cache: bool = False,
        padding_mask: Optional[torch.LongTensor] = None,
```

I didn't follow the Llama implementation too closely but why do we have both an `attention_mask` and a `padding_mask` and what is the difference here? The masking code is getting pretty confusing here

Reply:
Thank you. This additional argument was introduced in https://github.com/huggingface/transformers/pull/25598 and is essentially used to detect whether any of the token is masked. For FA2, it is useful to know whether to call `flash_attn_varlen_func` (at least one token is masked), or `flash_attn_func` (no token masked - faster). Maybe `masking_mask` would be a better name :smile: 

Here `padding_mask` is used to detect if we can potentially drop the Transformers `attention_mask`, and instead rely on PyTorch to generate a mask if needed. This is especially useful for batch_size = 1 passing `attn_mask=None`, where SDPA dispatcher will be able to dispatch to FA (or to FA2 in nightly).

Reply:
I see, ok this is getting a bit complex. But in essence we don't really need `padding_mask` here functionally no? It's only used to figure out whether we have "no" attention_mask and can thus dispatch for FA no?

Reply:
Fixed in your attention mask refactor